File: app/src/main/java/app/fit/fitndflow/ui/features/training/AddSerieTrainingFragment.java
package app.fit.fitndflow.ui.features.training;

import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModelProviders;
import androidx.recyclerview.widget.LinearLayoutManager;

import com.fit.fitndflow.R;
import com.fit.fitndflow.databinding.AddSerieTrainingFragmentBinding;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

import app.fit.fitndflow.domain.model.ExerciseModel;
import app.fit.fitndflow.domain.model.SerieModel;
import app.fit.fitndflow.ui.features.categories.ConfirmationDialogFragment;
import app.fit.fitndflow.ui.features.categories.DialogCallbackDelete;
import app.fit.fitndflow.ui.features.common.CommonFragment;
import app.fit.fitndflow.ui.features.home.HomeViewModel;

public class AddSerieTrainingFragment extends CommonFragment implements TrainingCallback, DialogCallbackDelete {

    public static final double INPUT_KG = 2.5;
    public static final int INPUT_REPS = 1;
    public static final int INPUT_ZERO = 0;
    public static final String KEY_EXCERCISE = "actualExercise";
    private AddSerieTrainingFragmentBinding binding;

    private HomeViewModel homeViewModel;
    private ExerciseModel exercise;
    private SeriesAdapter seriesAdapter;

    private SerieModel serieModel;


    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        binding = AddSerieTrainingFragmentBinding.inflate(getLayoutInflater());
        View view = binding.getRoot();
        super.onCreateView(inflater, container, savedInstanceState);

        Bundle bundle = getArguments();
        if(bundle != null){
            exercise = (ExerciseModel) bundle.getSerializable(KEY_EXCERCISE);
        }
        binding.exerciseNameTitle.setText(exercise.getName());
        homeViewModel = ViewModelProviders.of(getActivity()).get(HomeViewModel.class);
        instantiateSeriesAdapter();
        setViewModelObservers();
        initListeners();
        return view;
    }

    @Override
    protected Class getViewModelClass() {
        return HomeViewModel.class;
    }

    public static AddSerieTrainingFragment newInstance(ExerciseModel exercise){
        AddSerieTrainingFragment addSerieTrainingFragment = new AddSerieTrainingFragment();
        Bundle bundle = new Bundle();
        bundle.putSerializable(KEY_EXCERCISE, exercise);
        addSerieTrainingFragment.setArguments(bundle);
        return addSerieTrainingFragment;
    }
    private void setViewModelObservers(){
        final Observer<HashMap<Integer, List<SerieModel>>> observer = new Observer<HashMap<Integer, List<SerieModel>>>() {
            @Override
            public void onChanged(HashMap<Integer, List<SerieModel>> actualHashMap) {
                seriesAdapter.setSerieModelList(actualHashMap.get(exercise.getId()));
            }
        };
        homeViewModel.getHashmapMutableSerieListByExerciseId().observe(getActivity(), observer);

        final Observer<Boolean> observerIsSaveSuccess = new Observer<Boolean>() {
            @Override
            public void onChanged(Boolean isSaveSuccess) {
                if (isSaveSuccess) {
                    showSlideSaved();
                    homeViewModel.getIsSaveSuccess().setValue(false);
                }
            }
        };
        homeViewModel.getIsSaveSuccess().observe(getActivity(), observerIsSaveSuccess);

        final Observer<Boolean> observerError = new Observer<Boolean>() {
            @Override
            public void onChanged(Boolean isError) {
                if(isError){
                    showSlideError();
                    homeViewModel.getMutableSlideError().setValue(false);
                }
            }
        };
        homeViewModel.getMutableSlideError().observe(getActivity(), observerError);

        final Observer<Boolean> errorObserver = new Observer<Boolean>() {
            @Override
            public void onChanged(Boolean isError) {
                if (isError) {
                    printError();
                    homeViewModel.getMutableFullScreenError().setValue(false);
                }
            }
        };
        homeViewModel.getMutableFullScreenError().observe(getActivity(), errorObserver);

        final Observer<Boolean> observerLoading = new Observer<Boolean>() {
            @Override
            public void onChanged(Boolean isLoading) {
                try {
                    if (isLoading) {
                        showLoading();

                    } else {
                        hideLoading();
                    }
                } catch (Exception exception) {
                    Log.e("Error", "show loading");
                }
            }
        };
        homeViewModel.getIsLoading().observe(getActivity(), observerLoading);
    }

    private void initListeners(){
        binding.saveAndUpdateBtn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                int reps = binding.etCounterReps.getText().toString().equals("")? 0 : Integer.parseInt(binding.etCounterReps.getText().toString());
                double kg = binding.etCounterKg.getText().toString().equals("")? 0 : Double.parseDouble(binding.etCounterKg.getText().toString());
                if(serieModel != null){
                   homeViewModel.modifySerie(requireContext(),serieModel.getId(), reps, kg, exercise.getId());
                } else {
                    //utilizamos el operador ternario como un if: texto es igual(equals) a vacío ("")? (entonces pinta) 0 : (sino) pinta texto

                    homeViewModel.addNewSerie(requireContext(), reps, kg, exercise.getId());
                }
            }
        });

        binding.deleteAndCleanBtn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                if(serieModel != null){
                    showDeleteDialog(serieModel.getId());
                } else {
                    binding.etCounterReps.setText("");
                    binding.etCounterKg.setText("");
                }
            }
        });

        binding.btnIncrement.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                if(binding.etCounterKg.getText().toString().equals("")) {
                    binding.etCounterKg.setText(Double.toString(INPUT_KG));
                } else {
                    double currentValue = Double.parseDouble(binding.etCounterKg.getText().toString());
                    double newValue = currentValue + INPUT_KG;
                    binding.etCounterKg.setText(Double.toString(newValue));
                }
            }
        });
        binding.btnDecrement.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                if(binding.etCounterKg.getText().toString().equals("")) {
                    binding.etCounterKg.setText(Double.toString(INPUT_ZERO));
                } else {
                    double currentValue = Double.parseDouble(binding.etCounterKg.getText().toString());
                    double newValue = currentValue - INPUT_KG;
                    if(newValue > INPUT_ZERO) {
                        binding.etCounterKg.setText(Double.toString(newValue));
                    } else{
                        binding.etCounterKg.setText(Double.toString(INPUT_ZERO));
                    }
                }
            }
        });

        binding.btnIncrement2.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                if(binding.etCounterReps.getText().toString().equals("")) {
                    binding.etCounterReps.setText(Integer.toString(INPUT_REPS));
                } else {
                    int currentValue = Integer.valueOf(binding.etCounterReps.getText().toString());
                    int newValue = currentValue + INPUT_REPS;
                    binding.etCounterReps.setText(Integer.toString(newValue));
                }
            }
        });

        binding.btnDecrement2.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                if(binding.etCounterReps.getText().toString().equals("")) {
                    binding.etCounterReps.setText(Integer.toString(INPUT_ZERO));
                } else {
                    int currentValue = Integer.valueOf(binding.etCounterReps.getText().toString());
                    int newValue = currentValue - INPUT_REPS;
                    if(newValue > INPUT_ZERO) {
                        binding.etCounterReps.setText(Integer.toString(newValue));
                    } else {
                        binding.etCounterReps.setText(Integer.toString(INPUT_ZERO));
                    }
                }
            }
        });
    }

    private void instantiateSeriesAdapter(){
        seriesAdapter = new SeriesAdapter(new ArrayList<>(), this);
        binding.addSerieLayout.setHasFixedSize(true);
        binding.addSerieLayout.setLayoutManager(new LinearLayoutManager(this.getContext()));
        binding.addSerieLayout.setAdapter(seriesAdapter);
    }

    @Override
    public void clickListenerInterfaceAdapter(SerieModel input) {
        serieModel = input;
        if(input != null){
            binding.etCounterReps.setText(input.getReps() != null? Integer.toString(input.getReps()) : "0");
            binding.etCounterKg.setText(input.getKg() != null? Double.toString(input.getKg()) : "0");
            binding.saveAndUpdateBtn.setText(R.string.update);
            binding.deleteAndCleanBtn.setText(R.string.delete_btn);
        } else {
            binding.etCounterReps.setText("");
            binding.etCounterKg.setText("");
            binding.saveAndUpdateBtn.setText(R.string.add);
            binding.deleteAndCleanBtn.setText(R.string.clean);
        }

    }
    private void printError() {
        try {
            showBlockError();
        } catch (Exception exception) {
            Log.e("Error", "Error to print errorContainer");
        }
    }
    private void showDeleteDialog(int id) {
        ConfirmationDialogFragment.newInstance(AddSerieTrainingFragment.this, ConfirmationDialogFragment.DELETE_SERIE, id).show(getChildFragmentManager(), "ConfirmationDialog");
    }

    @Override
    public void onClickAcceptDelete(int id) {
        Toast.makeText(requireContext(), "delete serie", Toast.LENGTH_SHORT).show();
        //todo borrar serie del listado de series
    }
}

File: app/src/main/java/app/fit/fitndflow/ui/features/training/AddSerieTrainingFragment.kt
package app.fit.fitndflow.ui.features.training

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.viewModels
import androidx.lifecycle.lifecycleScope
import androidx.recyclerview.widget.LinearLayoutManager
import app.fit.fitndflow.domain.model.ExerciseModel
import app.fit.fitndflow.domain.model.SerieModel
import app.fit.fitndflow.ui.features.categories.ConfirmationDialogFragment
import app.fit.fitndflow.ui.features.categories.DialogCallbackDelete
import app.fit.fitndflow.ui.features.common.CommonFragment
import app.fit.fitndflow.ui.features.common.getTranslatedString
import app.fit.fitndflow.ui.features.common.hideKeyBoard
import com.fit.fitndflow.R
import com.fit.fitndflow.databinding.AddSerieTrainingFragmentBinding
import dagger.hilt.android.AndroidEntryPoint
import kotlinx.coroutines.flow.launchIn
import kotlinx.coroutines.flow.onEach

@AndroidEntryPoint
class AddSerieTrainingFragment : CommonFragment(), TrainingCallback, DialogCallbackDelete {

    companion object {
        val INPUT_KG = 2.5
        val INPUT_REPS = 1
        val INPUT_ZERO = 0
        val KEY_EXERCISE = "actualExercise"

        @JvmStatic
        fun newInstance(exercise: ExerciseModel): AddSerieTrainingFragment {
            val addSerieTrainingFragment = AddSerieTrainingFragment()
            val bundle = Bundle()
            bundle.putSerializable(KEY_EXERCISE, exercise)
            addSerieTrainingFragment.arguments = bundle
            return addSerieTrainingFragment
        }
    }

    private lateinit var binding: AddSerieTrainingFragmentBinding
    private val addSerieTrainingViewModel: AddSerieTrainingViewModel by viewModels()
    private val exercise: ExerciseModel by lazy { requireArguments().getSerializable(KEY_EXERCISE) as ExerciseModel }
    private lateinit var seriesAdapter: SeriesAdapter
    private var currentSelectedSerieModel: SerieModel? = null
    private var lastRepsWritten: Int? = null
    private var lastKgWritten: Double? = null

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        binding = AddSerieTrainingFragmentBinding.inflate(layoutInflater)
        val myView = binding.root
        super.onCreateView(inflater, container, savedInstanceState)
        binding.exerciseNameTitle.setText(exercise.name.getTranslatedString(requireContext()))
        initListeners()
        return myView
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        attachObservers()
        addSerieTrainingViewModel.getSerieListOfExerciseAdded(exercise.id!!)
    }

    private fun attachObservers() {
        addSerieTrainingViewModel.state.onEach(::handleState)
            .launchIn(viewLifecycleOwner.lifecycleScope)
    }

    private fun handleState(state: AddSerieTrainingViewModel.State) {
        when (state) {
            is AddSerieTrainingViewModel.State.SeriesChangedInExerciseDetail -> {
                exercise.record = state.serieInfoWrapper.newRecord
                instantiateSeriesAdapter(state.serieInfoWrapper.serieList)
                setScreenEditMode(false)
                showSlideSaved()
                if (state.showLastSerieAdded) {
                    binding.apply{
                        etCounterReps.setText(lastRepsWritten.toString())
                        etCounterKg.setText(lastKgWritten.toString())
                    }
                }
                if(state.serieInfoWrapper.showRecord){
                    showRecordDialog()
                }

                hideLoading()
            }
            is AddSerieTrainingViewModel.State.SerieListRecived -> {
                instantiateSeriesAdapter(state.serieList)
                setScreenEditMode(false)
                printFirstSerieAdded()
                hideLoading()
            }

            AddSerieTrainingViewModel.State.FullScreenError -> {
                hideLoading()
                showBlockError()
            }

            AddSerieTrainingViewModel.State.SlideError -> {
                hideLoading()
                showSlideError()
            }

            AddSerieTrainingViewModel.State.Loading -> {
                showLoading()
            }
        }
    }

    private fun initListeners() {
        binding.apply {
            saveAndUpdateBtn.setOnClickListener {
                val reps: Int = if (etCounterReps.text.toString()
                        .isEmpty()
                ) 0 else etCounterReps.text.toString().toInt()
                lastRepsWritten = reps

                val kg: Double = if (etCounterKg.text.toString()
                        .isEmpty()
                ) 0.0 else etCounterKg.text.toString().toDouble()
                lastKgWritten = kg

                currentSelectedSerieModel?.let {
                    addSerieTrainingViewModel.modifySerie(it.id!!, reps, kg, exercise.record)
                } ?: addSerieTrainingViewModel.addNewSerie(reps, kg, exercise.id!!, exercise.record)
                activity?.hideKeyBoard()
            }

            deleteAndCleanBtn.setOnClickListener {
                if (currentSelectedSerieModel != null) {
                    showDeleteDialog(currentSelectedSerieModel!!.id!!)
                } else {
                    etCounterReps.setText("")
                    etCounterKg.setText("")
                }
            }

            btnIncrement.setOnClickListener {
                val currentValue: Double = if (etCounterKg.text.toString()
                        .isEmpty()
                ) 0.0 else etCounterKg.text.toString().toDouble()
                val newValue = currentValue + INPUT_KG
                etCounterKg.setText(newValue.toString())
            }

            btnDecrement.setOnClickListener {
                val currentValue =
                    if (etCounterKg.text.toString().isEmpty()) 0.0 else etCounterKg.text.toString()
                        .toDouble()
                val newValue = currentValue - INPUT_KG
                if (newValue > INPUT_ZERO) {
                    etCounterKg.setText(newValue.toString())
                } else {
                    etCounterKg.setText(INPUT_ZERO.toString())
                }
            }

            btnIncrement2.setOnClickListener {
                val currentValue = if (etCounterReps.text.toString()
                        .isEmpty()
                ) 0 else etCounterReps.text.toString().toInt()
                val newValue = currentValue + INPUT_REPS
                etCounterReps.setText(newValue.toString())
            }

            btnDecrement2.setOnClickListener {
                val currentValue = if (etCounterReps.text.toString()
                        .isEmpty()
                ) 0 else etCounterReps.text.toString().toInt()
                val newValue = currentValue - INPUT_REPS
                if (newValue > INPUT_ZERO) {
                    etCounterReps.setText(newValue.toString())
                } else {
                    etCounterReps.setText(INPUT_ZERO.toString())
                }
            }
        }
    }

    private fun instantiateSeriesAdapter(serieModelList: List<SerieModel>) {
        seriesAdapter = SeriesAdapter(serieModelList, this)
        binding.apply {
            addSerieLayout.setHasFixedSize(true)
            addSerieLayout.layoutManager = LinearLayoutManager(requireContext())
            addSerieLayout.adapter = seriesAdapter
        }
    }

    private fun setScreenEditMode(isEditMode: Boolean) {
        binding.apply {
            if (isEditMode) {
                etCounterReps.setText(currentSelectedSerieModel!!.reps?.toString() ?: "0")
                etCounterKg.setText(currentSelectedSerieModel!!.kg?.toString() ?: "0.0")
                saveAndUpdateBtn.setText(R.string.update)
                saveAndUpdateBtn.setBackgroundResource(R.drawable.shape_serie_add_btn)
                deleteAndCleanBtn.setText(R.string.delete_btn)
                deleteAndCleanBtn.setBackgroundResource(R.drawable.shape_serie_delete_btn)
            } else {
                currentSelectedSerieModel = null
                etCounterReps.setText("")
                etCounterKg.setText("")
                saveAndUpdateBtn.setText(R.string.add)
                saveAndUpdateBtn.setBackgroundResource(R.drawable.shape_serie_add_and_clean_btns)
                deleteAndCleanBtn.setText(R.string.clean)
                deleteAndCleanBtn.setBackgroundResource(R.drawable.shape_serie_add_and_clean_btns)
            }
        }
    }
    private fun printFirstSerieAdded(){
        exercise.lastFirstSerie?.let {
            lastRepsWritten = if (exercise.lastFirstSerie!!.reps == null) 0 else exercise.lastFirstSerie!!.reps
            lastKgWritten = if (exercise.lastFirstSerie!!.kg == null) 0.0 else exercise.lastFirstSerie!!.kg
            binding.apply {
                etCounterReps.setText(lastRepsWritten.toString())
                etCounterKg.setText(lastKgWritten.toString())
            }
        }
    }

    private fun showRecordDialog(){
        exercise.record?.let {
            RecordSerieDialog.newInstance().show(childFragmentManager, RecordSerieDialog.TAG)
        }
    }

    private fun showDeleteDialog(id: Int) {
        ConfirmationDialogFragment.newInstance(this, ConfirmationDialogFragment.DELETE_SERIE, id)
            .show(childFragmentManager, ConfirmationDialogFragment.TAG)
    }

    override fun onClickAcceptDelete(serieId: Int) {
        addSerieTrainingViewModel.deleteSerie(serieId)
    }

    override fun clickListenerInterfaceAdapter(input: SerieModel?) {
        currentSelectedSerieModel = input
        val isEditMode = input != null
        setScreenEditMode(isEditMode)
    }
}
